### DISCUSSION TITLE
Why IntelliSense triggers on hover in comments?

### DISCUSSION THREAD

**Discussion by xHasKx** [microsoft/vscode-cpptools#8803] [Q&A] [answered]
Hi, I'm trying to debug why there is an IntelliSense tooltip that appears when I just hover my mouse over code comments.
Often the word in the comment is commonly used, which leads to high CPU/IO load and useless tooltip.
Sometimes VSCode is even stuck in that process when I stop the mouse cursor in an unlucky place, which is very annoying.
Can somebody help me to figure out how to disable that behavior?

![image](https://user-images.githubusercontent.com/1585990/152988908-f20420ea-16bc-4f98-b627-1f08b116ff35.png)

**Comment by sean-mcmanus** [comment, marked as answer]
The hover in the comment should only occur if IntelliSense is not used and it's falling back to the "Tag Parser" for some reason. Is your intelliSenseMode set to "Default" or "Tag Parser"? Does the file have any missing includes?

**Reply by xHasKx** [reply]
Thanks, I've checked my config, I have "C_Cpp.intelliSenseEngine": "Default" and "C_Cpp.default.intelliSenseMode": "gcc-x64" set. I didn't know about Tag Parser fallback.

I've also found setting "C_Cpp.intelliSenseEngineFallback", switched it to "Enabled" and then to "Disabled" for sure. But still, there is a tooltip appearing on hover in a comment in that file. Maybe it's not related to my issue?

I have also checked my file and didn't find any visible include issues.
I also noted that the tooltip in the comment appears only after a few minutes after the VSCode window reload. It's possible that "broken includes" can be somewhere in other files in the project (but the project is compiling properly). Is there an easy way to find them?
I checked all dropdown items like "C/C++ Configuration Warnings" at the bottom right on the "OUTPUT" tab but didn't find anything.

**Reply by sean-mcmanus** [reply]
Hover will fallback to Tag Parser results if IntelliSense isn't ready yet (you can see this in the C/C++ logging if C_Cpp.loggingLevel is "Debug"). You need to wait for IntelliSense to finish it's initial initialization (first flame icon disappears).

**Reply by xHasKx** [reply]
It seems I resolved my issue. After setting C_Cpp.loggingLevel: Debug I found lines reporting about constantly changed files causing IntelliSense re-generation every few seconds (there is a postgres database running inside the project's folder):

`cpptools/fileCreated: file:///home/.../run/postgres/pg_data/pg_stat_tmp/db_0.stat`

Then I added that `run/postgres` path to the `files.watcherExclude` workspace setting, reloaded window, and now can't reproduce tooltip over comments.

It seems that if IntelliSense scans a lot of project files (`Discovering files: 929124 file(s) processed`) over and over, then after a few minutes something is breaking inside it which causes fallback to Tag Parser causing tooltip on comments.

Thanks @sean-mcmanus a lot for your help!

**Comment by tlc** [comment]
@sean-mcmanus,
I'm curious about what this question might imply:  "**Does the file have any missing includes?**"

I'm currently working on porting part of a LARGE code base.  There are often missing headers at any given moment, as I decide what files include.

I'm also chasing CPU issues in cpptools I'm only starting to debug.  CPU usage is high for cpptools and vs code is very slow to change the problem list when files are edited and saved or added to the module.

Is it known that cpptools behaves poorly when not everything needed is found?  Could this cause looping?

**Reply by sean-mcmanus** [reply]
@tlc If there are missing includes, then it'll fallback to tag parser mode if "C_Cpp.intelliSenseEngineFallback": "enabled" is set, but that is not the default value for that setting. It's certainly possible that there could be a performance issue or looping problem if includes are not found, since it'll cause our parser to potentially try to parse code it doesn't understand, but it's not a widespread/known issue, i.e. the known bugs have been fixed. You can set C_Cpp.loggingLevel to "Debug" to get more info on what is happening, but if that's not enough you can attach a debugger to the cpptools or cpptools-srv process and break into the debugger to get a call stack: https://github.com/microsoft/vscode-cpptools/wiki/Attaching-debugger-to-cpptools-or-cpptools%E2%80%90srv .
